samplesheets/assayapps/generic/plugins.py

```diff
@@ -0,0 +1,206 @@
+        :param row: List of dicts (a row returned by SampleSheetTableBuilder)
+        :param table: Full table with headers (dict returned by
+                      SampleSheetTableBuilder)
+        :param assay: Assay object
+        :param assay_path: Root path for assay
+        :return: String with full iRODS path or None
+        """
+        # Extract comments starting with DATA_COMMENT_PREFIX; sorted
+        data_columns = [
+            value
+            for name, value in sorted(assay.comments.items())
+            if name.startswith(DATA_COMMENT_PREFIX)
+        ]
+
+        data_collections = []
+        for column_name in data_columns:
+            col_value = self._get_col_value(column_name, row, table)
+            if col_value:
+                data_collections.append(col_value)
+
+        # Build iRODS path from list and stop at first None value
+        if data_collections:
+            data_path = '/' + '/'.join(data_collections)
+            return assay_path + data_path
+        return None
+
+    def update_row(self, row, table, assay, index):
+        """
+        Update render table row with e.g. links. Return the modified row.
+
+        :param row: Original row (list of dicts)
+        :param table: Full table (dict)
+        :param assay: Assay object
+        :param index: Row index (int)
+        :return: List of dicts
+        """
+        if not settings.IRODS_WEBDAV_ENABLED or not assay:
+            return row
+        assay_path = self.get_assay_path(assay)
+        if not assay_path:
+            return row
+
+        base_url = settings.IRODS_WEBDAV_URL + assay_path
+        top_header = None
+        th_colspan = 0
+
+        results_cols = assay.comments.get(RESULTS_COMMENT)
+        if results_cols:
+            results_cols = results_cols.lower().split(';')
+        misc_cols = assay.comments.get(MISC_FILES_COMMENT)
+        if misc_cols:
+            misc_cols = misc_cols.lower().split(';')
+        data_cols = assay.comments.get(DATA_LINK_COMMENT)
+        if data_cols:
+            data_cols = data_cols.lower().split(';')
+            if table['irods_paths'][index]:
+                row_path = table['irods_paths'][index]['path']
+            else:
+                row_path = self.get_row_path(row, table, assay, assay_path)
+
+        for i in range(len(row)):
+            header = table['field_header'][i]
+            if not top_header or i >= th_colspan:
+                top_header = get_top_header(table, i)
+                th_colspan += top_header['colspan']
+
+            # TODO: Check if two comments reference the same column header?
+            # Create Results links
+            if results_cols:
+                self._link_from_comment(
+                    row[i],
+                    header,
+                    top_header,
+                    results_cols,
+                    f'{base_url}/{RESULTS_COLL}',
+                )
+            # Create MiscFiles links
+            if misc_cols:
```

Comment:
Here we still have the problem that we attempt to modify the same field up to three times, even if we already did.

Also, if the user has specified multiple collection types for linking and added the same field names (accidentally?) into many collection types, we may modify the same cell value multiple times and probably malign the resulting link in the process.

My suggestion: have `_link_from_comment()` return e.g. `True` if the value was modified. If it was, add a `continue` after the fuction call and skip the rest.

In a perfect word, we should also warn the user that they have configured the plugin in a way where they attempt to link a column for multiple different collections. However, at the moment there is no callback to e.g. Django messages in these assay plugins, so the warning would only be visible in the system logs. I'll have to think about how to add a warning mechanism to these plugins in the future.

Reply:
done